File: server/lib/agent/orchestrator.dart
import 'dart:async';
import 'dart:convert';

import 'package:logging/logging.dart';
import 'package:uuid/uuid.dart';

import '../db/repositories/agent_runs_repo.dart';
import '../db/repositories/work_items_repo.dart';
import '../ollama/ollama_client.dart';
import 'executor.dart';
import 'planner.dart';
import 'reflector.dart';
import 'tool_registry.dart';
import 'tools/tool.dart';

class Orchestrator {
  Orchestrator({
    required this.workItems,
    required this.runs,
    required this.tools,
    required this.ollama,
    required this.maxSteps,
  })  : _planner = Planner(ollama),
        _executor = Executor(ollama: ollama, tools: tools),
        _reflector = Reflector(ollama);

  final WorkItemsRepository workItems;
  final AgentRunsRepository runs;
  final ToolRegistry tools;
  final OllamaClient ollama;
  final int maxSteps;

  final Planner _planner;
  final Executor _executor;
  final Reflector _reflector;
  final Logger _log = Logger('Orchestrator');
  final Uuid _uuid = const Uuid();

  /// Kicks off a run. Returns the new run id immediately; the run continues
  /// in the background. Subscribers can poll `/api/runs/<id>`.
  Future<String> startRun(String workItemId) async {
    final item = await workItems.get(workItemId);
    if (item == null) {
      throw ArgumentError('work item $workItemId not found');
    }
    final runId = _uuid.v4();
    await runs.create(
      id: runId,
      workItemId: workItemId,
      model: ollama.model,
      maxSteps: maxSteps,
    );

    // Run async; surface failures into the run record.
    unawaited(_drive(runId, workItemId));
    return runId;
  }

  Future<void> _drive(String runId, String workItemId) async {
    try {
      await runs.setStatus(runId, 'running');
      var item = (await workItems.get(workItemId))!;

      final history = <OllamaMessage>[];

      var plan = await _planner.plan(item);
      await _persistPlan(runId, plan);
      _log.info('run $runId planned ${plan.length} steps');

      var executed = 0;
      var ordinal = 0;
      while (executed < maxSteps && ordinal < plan.length) {
        final stored = await runs.listSteps(runId);
        final pending = stored.firstWhere(
          (s) => s.status == 'pending',
          orElse: () => stored.last,
        );
        await runs.updateStep(id: pending.id, status: 'running', markStarted: true);

        final ctx = ToolContext(workItemId: workItemId, runId: runId);
        final outcome = await _executor.execute(
          item: item,
          stepTitle: pending.title,
          stepRationale: pending.rationale,
          history: history,
          ctx: ctx,
        );

        await _recordOutcome(runId, pending.id, outcome, history);
        await runs.updateStep(
          id: pending.id,
          status: 'done',
          result: _summariseOutcome(outcome),
          markFinished: true,
        );

        executed++;
        ordinal++;

        final reflection = await _reflector.reflect(
          history: history,
          latestStepTitle: pending.title,
          latestStepOutcome: _summariseOutcome(outcome),
        );
        await runs.addMessage(
          id: _uuid.v4(),
          runId: runId,
          stepId: pending.id,
          role: 'system',
          content: 'reflection: ${reflection.decision.name} — ${reflection.note}',
        );

        switch (reflection.decision) {
          case ReflectionDecision.continueNext:
            break;
          case ReflectionDecision.replan:
            item = (await workItems.get(workItemId))!;
            plan = await _planner.plan(item);
            await _persistPlan(runId, plan, replan: true);
            ordinal = 0;
            break;
          case ReflectionDecision.done:
            await runs.setSummary(runId, reflection.summary ?? reflection.note);
            await workItems.updateState(workItemId, 'Resolved');
            await runs.setStatus(runId, 'done');
            return;
          case ReflectionDecision.blocked:
            await runs.setSummary(
              runId,
              'Blocked: ${reflection.note}',
            );
            await runs.setStatus(runId, 'blocked');
            return;
        }
      }

      await runs.setSummary(
        runId,
        'Reached max steps ($maxSteps) without converging.',
      );
      await runs.setStatus(runId, 'blocked');
    } catch (e, st) {
      _log.severe('run $runId failed', e, st);
      await runs.setStatus(runId, 'failed', error: '$e\n$st');
    }
  }

  Future<void> _persistPlan(
    String runId,
    List<PlannedStep> plan, {
    bool replan = false,
  }) async {
    if (replan) {
      // We don't delete prior steps - they remain as history; we just append
      // the new ones after the highest existing ordinal.
      final existing = await runs.listSteps(runId);
      final base = existing.isEmpty ? 0 : existing.last.ordinal + 1;
      for (var i = 0; i < plan.length; i++) {
        await runs.addStep(
          id: _uuid.v4(),
          runId: runId,
          ordinal: base + i,
          title: plan[i].title,
          rationale: plan[i].rationale,
        );
      }
    } else {
      for (var i = 0; i < plan.length; i++) {
        await runs.addStep(
          id: _uuid.v4(),
          runId: runId,
          ordinal: i,
          title: plan[i].title,
          rationale: plan[i].rationale,
        );
      }
    }
  }

  Future<void> _recordOutcome(
    String runId,
    String stepId,
    StepOutcome outcome,
    List<OllamaMessage> history,
  ) async {
    if (outcome.assistantContent.isNotEmpty) {
      await runs.addMessage(
        id: _uuid.v4(),
        runId: runId,
        stepId: stepId,
        role: 'assistant',
        content: outcome.assistantContent,
      );
      history.add(
        OllamaMessage(role: 'assistant', content: outcome.assistantContent),
      );
    }
    for (var i = 0; i < outcome.toolCalls.length; i++) {
      final call = outcome.toolCalls[i];
      final res = outcome.toolResults[i];
      final argsJson = jsonEncode(call.arguments);
      final resJson = jsonEncode(res.result);
      await runs.addMessage(
        id: _uuid.v4(),
        runId: runId,
        stepId: stepId,
        role: 'tool',
        content: 'call ${call.name}($argsJson) -> $resJson',
        toolName: call.name,
      );
      await runs.recordToolCall(
        id: _uuid.v4(),
        runId: runId,
        stepId: stepId,
        toolName: call.name,
        argumentsJson: argsJson,
        resultJson: resJson,
        success: res.success,
      );
      history.add(
        OllamaMessage(
          role: 'tool',
          toolName: call.name,
          content: resJson,
        ),
      );
    }
  }

  String _summariseOutcome(StepOutcome o) {
    if (o.toolResults.isEmpty) return o.assistantContent;
    final parts = <String>[
      if (o.assistantContent.isNotEmpty) o.assistantContent,
      ...o.toolResults.map(
        (r) => 'tool ${r.name}: ${r.success ? 'ok' : 'error'} '
            '${jsonEncode(r.result)}',
      ),
    ];
    return parts.join('\n');
  }
}

File: server/app/agent/orchestrator.py
from __future__ import annotations

import asyncio
import json
import logging
import uuid

from ..db.repositories.agent_runs import AgentRunsRepository
from ..db.repositories.work_items import WorkItemsRepository
from .executor import Executor, StepOutcome
from .ollama_client import OllamaClient, OllamaMessage
from .planner import PlannedStep, Planner
from .reflector import Reflection, ReflectionDecision, Reflector
from .tools.base import ToolContext, ToolRegistry

log = logging.getLogger(__name__)


class Orchestrator:
    def __init__(
        self,
        *,
        work_items: WorkItemsRepository,
        runs: AgentRunsRepository,
        tools: ToolRegistry,
        ollama: OllamaClient,
        max_steps: int,
    ) -> None:
        self.work_items = work_items
        self.runs = runs
        self.tools = tools
        self.ollama = ollama
        self.max_steps = max_steps
        self._planner = Planner(ollama)
        self._executor = Executor(ollama, tools)
        self._reflector = Reflector(ollama)

    async def start_run(self, work_item_id: str) -> str:
        item = await self.work_items.get(work_item_id)
        if item is None:
            raise ValueError(f"work item {work_item_id} not found")
        run_id = str(uuid.uuid4())
        await self.runs.create(
            id_=run_id,
            work_item_id=work_item_id,
            model=self.ollama.model,
            max_steps=self.max_steps,
        )
        # Fire and forget; failures surface via the run record.
        task = asyncio.create_task(self._drive(run_id, work_item_id))
        task.add_done_callback(lambda t: t.exception() and log.exception(
            "run task crashed", exc_info=t.exception()
        ))
        return run_id

    async def _drive(self, run_id: str, work_item_id: str) -> None:
        try:
            await self.runs.set_status(run_id, "running")
            item = await self.work_items.get(work_item_id)
            assert item is not None

            history: list[OllamaMessage] = []

            plan = await self._planner.plan(item)
            await self._persist_plan(run_id, plan, replan=False)
            log.info("run %s planned %d steps", run_id, len(plan))

            executed = 0
            ordinal = 0
            while executed < self.max_steps and ordinal < len(plan):
                stored = await self.runs.list_steps(run_id)
                pending = next(
                    (s for s in stored if s.status == "pending"),
                    stored[-1] if stored else None,
                )
                if pending is None:
                    break

                await self.runs.update_step(
                    id_=pending.id, status="running", mark_started=True
                )

                ctx = ToolContext(work_item_id=work_item_id, run_id=run_id)
                outcome = await self._executor.execute(
                    item=item,
                    step_title=pending.title,
                    step_rationale=pending.rationale,
                    history=history,
                    ctx=ctx,
                )

                await self._record_outcome(run_id, pending.id, outcome, history)
                await self.runs.update_step(
                    id_=pending.id,
                    status="done",
                    result=self._summarise_outcome(outcome),
                    mark_finished=True,
                )

                executed += 1
                ordinal += 1

                reflection = await self._reflector.reflect(
                    history=history,
                    latest_step_title=pending.title,
                    latest_step_outcome=self._summarise_outcome(outcome),
                )
                await self.runs.add_message(
                    id_=str(uuid.uuid4()),
                    run_id=run_id,
                    step_id=pending.id,
                    role="system",
                    content=f"reflection: {reflection.decision.value} — {reflection.note}",
                )

                if reflection.decision is ReflectionDecision.DONE:
                    await self.runs.set_summary(
                        run_id, reflection.summary or reflection.note
                    )
                    await self.work_items.update_state(work_item_id, "Resolved")
                    await self.runs.set_status(run_id, "done")
                    return
                if reflection.decision is ReflectionDecision.BLOCKED:
                    await self.runs.set_summary(run_id, f"Blocked: {reflection.note}")
                    await self.runs.set_status(run_id, "blocked")
                    return
                if reflection.decision is ReflectionDecision.REPLAN:
                    item = await self.work_items.get(work_item_id)
                    assert item is not None
                    plan = await self._planner.plan(item)
                    await self._persist_plan(run_id, plan, replan=True)
                    ordinal = 0

            await self.runs.set_summary(
                run_id, f"Reached max steps ({self.max_steps}) without converging."
            )
            await self.runs.set_status(run_id, "blocked")
        except Exception as e:  # noqa: BLE001
            log.exception("run %s failed", run_id)
            await self.runs.set_status(run_id, "failed", error=repr(e))

    async def _persist_plan(
        self, run_id: str, plan: list[PlannedStep], *, replan: bool
    ) -> None:
        if replan:
            existing = await self.runs.list_steps(run_id)
            base = existing[-1].ordinal + 1 if existing else 0
        else:
            base = 0
        for i, step in enumerate(plan):
            await self.runs.add_step(
                id_=str(uuid.uuid4()),
                run_id=run_id,
                ordinal=base + i,
                title=step.title,
                rationale=step.rationale,
            )

    async def _record_outcome(
        self,
        run_id: str,
        step_id: str,
        outcome: StepOutcome,
        history: list[OllamaMessage],
    ) -> None:
        if outcome.assistant_content:
            await self.runs.add_message(
                id_=str(uuid.uuid4()),
                run_id=run_id,
                step_id=step_id,
                role="assistant",
                content=outcome.assistant_content,
            )
            history.append(
                OllamaMessage(role="assistant", content=outcome.assistant_content)
            )
        for call, res in zip(outcome.tool_calls, outcome.tool_results, strict=False):
            args_json = json.dumps(call.arguments, default=str)
            res_json = json.dumps(res.result, default=str)
            await self.runs.add_message(
                id_=str(uuid.uuid4()),
                run_id=run_id,
                step_id=step_id,
                role="tool",
                content=f"call {call.name}({args_json}) -> {res_json}",
                tool_name=call.name,
            )
            await self.runs.record_tool_call(
                id_=str(uuid.uuid4()),
                run_id=run_id,
                step_id=step_id,
                tool_name=call.name,
                arguments_json=args_json,
                result_json=res_json,
                success=res.success,
            )
            history.append(
                OllamaMessage(role="tool", tool_name=call.name, content=res_json)
            )

    @staticmethod
    def _summarise_outcome(o: StepOutcome) -> str:
        if not o.tool_results:
            return o.assistant_content
        parts: list[str] = []
        if o.assistant_content:
            parts.append(o.assistant_content)
        for r in o.tool_results:
            status = "ok" if r.success else "error"
            parts.append(f"tool {r.name}: {status} {json.dumps(r.result, default=str)}")
        return "\n".join(parts)


# Local re-export for readability
Reflection = Reflection
